samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,639 @@
+		OPAE_MSG("Failed to unlock perf mutex");
+		goto err;
+	}
+	/* parse the format value */
+	ret = parse_Perf_Format(dev);
+	if (ret != FPGA_OK)
+		goto err;
+	/* parse the event value */
+	ret = parse_Perf_Event(dev);
+	if (ret != FPGA_OK)
+		goto err;
+
+	/* initialize the pea structure to 0 */
+	memset(&pea, 0, sizeof(struct perf_event_attr));
+	
+	if (opae_mutex_lock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to lock perf mutex");
+		goto err;
+	}
+
+	for (loop = 0; loop < g_fpga_perf->num_perf_events; loop++) {
+
+		if (g_fpga_perf->perf_events[0].fd <= 0)
+			grpfd = -1;
+		else
+			grpfd = g_fpga_perf->perf_events[0].fd;
+		
+		if (!g_fpga_perf->perf_events[loop].config)
+			continue;
+
+		pea.type = g_fpga_perf->type;
+		pea.size = sizeof(struct perf_event_attr);
+		pea.config = g_fpga_perf->perf_events[loop].config;
+		pea.disabled = 1;
+		pea.inherit = 1;
+		pea.sample_type = PERF_SAMPLE_IDENTIFIER;
+		pea.read_format = PERF_FORMAT_GROUP | PERF_FORMAT_ID;
+		fd = syscall(__NR_perf_event_open, &pea, -1, g_fpga_perf->cpumask, grpfd, 0);
+		if (fd == -1) {
+			OPAE_ERR("Error opening leader %llx\n", pea.config);
+			ret = FPGA_EXCEPTION;
+			goto out;
+		} else {
+			g_fpga_perf->perf_events[loop].fd = fd;
+			if (ioctl(g_fpga_perf->perf_events[loop].fd, PERF_EVENT_IOC_ID,
+						&g_fpga_perf->perf_events[loop].id) == -1) {
+				OPAE_ERR("PERF_EVENT_IOC_ID ioctl failed: %s",
+						strerror(errno));
+				ret = FPGA_EXCEPTION;
+				goto out;
+			}
+		}
+	}
+	if (ioctl(g_fpga_perf->perf_events[0].fd, PERF_EVENT_IOC_RESET,
+				PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_RESET ioctl failed: %s", strerror(errno));
+		ret = FPGA_EXCEPTION;
+		goto out;
+	}
+out:
+	udev_unref(udev);
+	udev_device_unref(dev);
+	opae_mutex_unlock(res, &fpga_perf_lock);
+	return ret;
+err:
+	udev_unref(udev);
+	udev_device_unref(dev);
+	return FPGA_EXCEPTION;
+}
+
+/* get fpga sbdf from token */
+fpga_result get_Fpga_Sbdf(fpga_token token,
+		uint16_t *segment,
+		uint8_t *bus,
+		uint8_t *device,
+		uint8_t *function)
+{
+	fpga_result res = FPGA_OK;
+	fpga_properties props = NULL;
+
+	if (!segment || !bus ||
+		!device || !function) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+	res = fpgaGetProperties(token, &props);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get properties");
+		return res;
+	}
+	res = fpgaPropertiesGetBus(props, bus);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get bus");
+		return res;
+	}
+	res = fpgaPropertiesGetSegment(props, segment);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Segment");
+		return res;
+	}
+	res = fpgaPropertiesGetDevice(props, device);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Device");
+		return res;
+	}
+	res = fpgaPropertiesGetFunction(props, function);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Function");
+		return res;
+	}
+
+	return res;
+}
+
+fpga_result fpgaPerfCounterEnum(fpga_token token)
```

Comment:
Does this function assume that  the device this token represents is on the same bdf as the fme? or does it assume that this token is a token for an FME?